File: arduino/tests/unit/test_frame_layer_offline.cpp
#include "../../include/types.h"
#include "../../include/frame_decoder.h"
#include "../../include/frame_encoder.h"
#include "../../include/crc16_ccitt_false.h"
#include "../../include/control_authority_controller.h"
#include <cstring>
#include <cstdio>

int main() {
    frame_decoder_reset();

    frame_t test_frame;
    uint8_t test_payload[] = {0x21, 0x34, 0xC5};

    frame_encode(&test_frame, msg_type_t::MSG_HEARTBEAT, test_payload, sizeof(test_payload));

    uint8_t buf[5 + FRAME_MAX_PAYLOAD];
    calc_serialized_buf(&test_frame, buf);

    for (int i = 0; i < 3 + test_frame.len + 2; i++) {
        frame_decoder_process_byte_cb(buf[i]);
    }

    printf("Frame Decoded: type: 0x%02X, len: %d, payload: ", (uint8_t)test_frame.type, test_frame.len);
    for (int i = 0; i < test_frame.len; i++) {
        printf("0x%02X, ", test_frame.payload[i]);
    }
    printf(" crc: 0x%04X", test_frame.crc);
}

File: pi5/tests/test_frame_layer_offline.py
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SRC = ROOT / "src"

sys.path.append(str(SRC))

import frame
import frame_encoder
import frame_decoder
import crc16_ccitt_false

test_frame = [0] * frame.FRAME_WIRE_SIZE

payload = bytes([0x21, 0x34, 0xC5])
frame_encoder.frame_encode(test_frame, frame.msg_type_t.MSG_HEARTBEAT, payload)

buf = [0] * (5 + frame.FRAME_MAX_PAYLOAD)
frame_encoder.calc_serialized_buf(test_frame, buf)

frame_decoder.frame_decoder_reset()

for b in buf[:3 + test_frame[frame.LEN_OFS] + 2]:
    frame_decoder.frame_decoder_process_byte(b)

crc = (test_frame[frame.CRC_OFS] << 8) | test_frame[frame.CRC_OFS + 1]
payload_hex = ' '.join(f'0x{b:02X}' for b in test_frame[frame.PAYLOAD_OFS : frame.PAYLOAD_OFS + test_frame[frame.LEN_OFS]])

print(f"Frame Received: type: 0x{test_frame[frame.MSG_TYPE_OFS]:02X}, payload = {payload_hex}, CRC: 0x{crc:04X}") 
